### ISSUE TITLE
JUnit4 to JUnit5 migration

### ISSUE THREAD

**Issue by UngureanuSerg** [apache/druid#13948]
Hi team,

A proposal to moving on after the adoption of java8+ is to start the migration from junit4 to jnuit5.

- JUnit 4 never advanced beyond Java 7, missing out on a lot of features from Java 8. JUnit 5 makes good use of the Java 8 features.
- A single jar library contains the entire framework. We need to import the whole library, even when we only require a particular feature. In JUnit 5, we get more granularity and can import only what's necessary.
- Only one test runner can execute tests at a time in JUnit 4 (e.g. @RunWith(Parameterized.class)
 or provide druid extension with the adoption of JUnit Platform). JUnit 5 allows multiple runners to work simultaneously.

As a strategy for the approach I would suggest to migrate one module at a time(it will be easier for the prs) and create tasks for the rest of the modules. As a starting point we can still depend on JUnit Vintage for backward compatibility with JUnit 4 but gradually moving to JUnit5.

I think all the point mentioned describes why it would be beneficial for apache druid as well as I would like to pick this issue if the rest of the community is interested.

**Comment by cryptoe** [comment]
@UngureanuSerg Thanks for starting this. Yes lets start the migration. Pretty excited  🚀 .
IMHO `JUnit Vintage` should be our last option. If we are able to move to the JUnit 5 using `Unit Platform` dependencies that would be event cooler.

**Comment by gianm** [comment]
> @UngureanuSerg Thanks for starting this. Yes lets start the migration. Pretty excited 🚀 .
> IMHO `JUnit Vintage` should be our last option. If we are able to move to the JUnit 5 using `Unit Platform` dependencies that would be event cooler.

I don't know the details, but if "JUnit Vintage" allows us to do a migration piece-by-piece instead of updating every test case at once, that sounds like a good thing. Updating every test case at once could potentially be a prohibitive amount of work for a single contributor.

**Comment by UngureanuSerg** [comment]
Hey Gian I didn't know that junit-vintage give this kind of capabilities but I will check it for sure

**Comment by dampcake** [comment]
In other projects I use junit-vintage to mix both JUnit5 and JUnit4 tests in the same suites. This allows for new tests or tests where you want to take advantage of the features to be converted while leaving all the existing tests on Junit4. Pretty much a best of both worlds scenario.

**Comment by kgyrtkirk** [comment]
there was a related PR to this: #16106 ; 

I would like to leave a note here: to make sure that future PRs upgrading to junit5 should also configure openrewrite via [rewrites.yml](https://github.com/apache/druid/blob/d982727a298526948961e470df0eade6344f1ea8/rewrite.yml#L28-L29) to ensure that no junit4 starts getting back into the upgraded testclasses (configuring that first and running the rewrite engine could jump-start the migration process as well)

**Comment by FrankChen021** [comment]
### SQL and MSQ migration order

For the SQL portion, [PR #19909](https://github.com/apache/druid/pull/19909) is the gate. We should not move on to the subsequent SQL-module migration until that PR has been merged. This ensures that later work builds on the migrated SQL leaf and Quidem test APIs.

Planned sequence:

1. **PR 1: SQL leaf + Quidem** — [#19909](https://github.com/apache/druid/pull/19909)
2. **PR 2: SQL core + schema** — after PR1
3. **PR 3: MSQ low-level** — after PR2
4. **PR 4: MSQ integration** — after PR2; this can be developed in parallel with PR3 if there is no file overlap
5. **PR 5: MSQ shared base + exec + POM cleanup** — after both PR3 and PR4

Dependency flow:

`#19909` → PR2 → (PR3 and PR4) → PR5

Each PR will be compiled locally before pushing. The affected tests, Maven Checkstyle, and applicable static checks will also be run. JUnit 4 and Hamcrest dependencies will be removed within each PR only after that PR completes the relevant source migration; a final repository-wide cleanup will follow the last migration.

**Comment by FrankChen021** [comment]
### Proposed batching for `processing`, benchmarks, and cloud-common JUnit 5 migration

The previous processing migration branch changed about 300 files. After reviewing its shared fixtures, inheritance relationships, and cross-module test dependencies, the migration can be rebuilt as four independent PRs of approximately 50–80 files each.

### Dependency diagram

```mermaid
flowchart TD
  P19875["#19875<br/>JupiterAssertions, TempDirExtension,<br/>and shared groupBy support"]
  Shared["Small shared prerequisite PR<br/>3–6 files<br/>Additive extensions and compatibility overloads"]

  B1["Batch 1 · ~74 files<br/>Common utilities, data, timeline,<br/>benchmarks, AWS/GCP common"]
  B2["Batch 2 · ~77 files<br/>Frame, expressions, nested/virtual columns,<br/>segment filters and joins"]
  B3["Batch 3 · ~56 files<br/>Segment core, serialization,<br/>incremental indexes"]
  B4["Batch 4 · ~73 files<br/>Aggregation, groupBy, scan, topN,<br/>timeseries, metadata and search"]

  ProcessingCleanup["Processing dependency cleanup<br/>Remove JUnit 4, Vintage, JUnitParams,<br/>migration-support and obsolete rules"]
  OtherModules["Remaining SQL, server, indexing-service,<br/>MSQ, embedded and extension migrations"]
  HamcrestCleanup["Repository-wide Hamcrest cleanup"]

  P19875 --> Shared
  Shared -->|shared assertions and stable fixtures| B1
  Shared -->|TempDir and logger extensions| B2
  Shared -->|CloserExtension| B3
  Shared -->|groupBy helper and stable APIs| B4

  B1 --> ProcessingCleanup
  B2 --> ProcessingCleanup
  B3 --> ProcessingCleanup
  B4 --> ProcessingCleanup

  ProcessingCleanup --> HamcrestCleanup
  OtherModules --> HamcrestCleanup
```

The four migration batches are siblings: none depends on another migration batch. Their only migration dependency is the small backward-compatible shared prerequisite.

First, create one intentionally small shared prerequisite PR from the latest `master`. It should add JUnit 5 adapters while preserving the JUnit 4 APIs still used by other modules:

- Add `CloserExtension` without changing or removing `CloserRule`.
- Add `LoggerCaptureExtension` without changing or removing `LoggerCaptureRule`.
- Add JUnit 5-compatible temp-directory overloads to `NestedDataTestUtils` while preserving its `TemporaryFolder` overloads.
- Keep reusable groupBy factory behavior in `GroupByQueryRunnerTestHelper` without renaming existing APIs.
- Reuse the `JupiterAssertions` and `TempDirExtension` support introduced by #19875.
- Do not remove JUnit 4, Vintage, or Hamcrest dependencies in this prerequisite.

After that prerequisite merges, create these sibling PRs directly from `master`. They will depend only on the shared prerequisite, not on one another:

### Proposed PR scopes

| PR | Scope | Approx. files | Dependency |
|---|---|---:|---|
| Shared prerequisite | Additive JUnit 5 fixture adapters and compatibility overloads | 3–6 | #19875 |
| Batch 1 (#19940 ) | Common Java utilities, collections, data input, HLL, Guice, timeline, `cloud/aws-common`, `cloud/gcp-common`, and benchmarks | 74 | Shared prerequisite only |
| Batch 2 (#19980 ) | Frame tests, expressions, expression macros, nested/virtual columns, segment filters, and joins | 77 | Shared prerequisite only |
| Batch 3 (#19981 ) | Segment core, serialization, incremental indexes, and the complete index-merger hierarchy | 56 | Shared prerequisite only |
| Batch 4 (#19982 ) | Aggregation, groupBy, scan, topN, timeseries, metadata, search, and remaining query tests | 73 | Shared prerequisite only |
| Processing dependency cleanup | Remove unused JUnit 4, Vintage, JUnitParams, migration-support, and obsolete rule dependencies | Small follow-up | Batches 1–4 |
| Repository-wide Hamcrest cleanup | Migrate shared matcher APIs and all remaining consumers before removing Hamcrest | Separate follow-up | Relevant repository migrations |

1. **Common utilities, data, timeline, benchmarks, and cloud common — about 74 files**
   - Common Java utilities, collections, data-input tests, HLL, Guice, and timeline tests.
   - `cloud/aws-common` and `cloud/gcp-common` tests.
   - Benchmark tests.
   - Keep the benchmark Hamcrest dependency while inherited SQL test APIs still expose Hamcrest types.

2. **Frame, expressions, nested data, segment filters, and joins — about 77 files**
   - Frame tests and `FrameTestUtil` consumers.
   - Expression tests, `MacroTestBase`, and all of its subclasses.
   - Nested-data and virtual-column tests.
   - Segment filter and join tests, including the `BaseHashJoinSegmentCursorFactoryTest` hierarchy.

3. **Segment core, serialization, and incremental indexes — about 56 files**
   - Segment core and column serialization tests.
   - Incremental-index tests.
   - `IndexMergerTestBase` and all of its subclasses.
   - Migrated users of the additive `CloserExtension`.

4. **Query tests — about 73 files**
   - Aggregation, groupBy, scan, topN, timeseries, metadata, search, and other query families.
   - Keep `GroupByQueryRunnerTest` and its closely related tests in this batch.
   - Preserve public/shared test APIs used by extensions, server, MSQ, and benchmarks.

During these four PRs, `processing/pom.xml` should retain JUnit 4, Vintage, and Hamcrest so that tests not yet migrated continue to compile. Shared Hamcrest APIs such as `DruidExceptionMatcher` and `ExceptionMatcher` should also remain unchanged because they are consumed by SQL, server, indexing-service, MSQ, embedded tests, and extensions.

After all four migration PRs merge:

- Create a focused processing dependency-cleanup PR to remove JUnit 4, Vintage, JUnitParams, migration-support, and obsolete rule dependencies that are no longer used.
- Handle Hamcrest removal separately by first migrating its shared APIs and repository-wide consumers; avoid introducing a duplicate in-project matcher framework solely to replace Hamcrest.

This structure keeps each main migration PR reviewable and independently compilable, while the only shared dependency is a small backward-compatible fixture PR.

**Comment by FrankChen021** [comment]
## Updated JUnit 5 migration plan

The plan was refreshed against the latest `master` (`cea0d99504`, August 17, 2026).

The following migration work is already merged:

- [#19875](https://github.com/apache/druid/pull/19875): selected `extensions-core` tests.
- [#19908](https://github.com/apache/druid/pull/19908): extensions-contrib tests plus shared processing `TestHelper` and group-by support.
- [#19909](https://github.com/apache/druid/pull/19909) and [#19939](https://github.com/apache/druid/pull/19939): SQL leaf/core migrations.
- [#19910](https://github.com/apache/druid/pull/19910), [#19920](https://github.com/apache/druid/pull/19920), [#19922](https://github.com/apache/druid/pull/19922), [#19923](https://github.com/apache/druid/pull/19923), [#19925](https://github.com/apache/druid/pull/19925), and [#19926](https://github.com/apache/druid/pull/19926): indexing-service/indexing migrations.
- [#19940](https://github.com/apache/druid/pull/19940), [#19981](https://github.com/apache/druid/pull/19981), and [#19982](https://github.com/apache/druid/pull/19982): processing migration batches and shared JUnit 5 extensions.

The original large server migration [#19911](https://github.com/apache/druid/pull/19911) was closed without merging. The latest `master` still contains the following JUnit 4 or Hamcrest source usage:

| Module | Source files |
|---|---:|
| `server` | 379 |
| `services` | 20 |
| `embedded-tests` | 17 |

### Remaining server migration batches

| PR | Scope | Approximate files |
|---|---|---:|
| PR1 ([#20031](https://github.com/apache/druid/pull/20031)) | Remaining shared server-fixture adaptations, client/query tests, and query-serving tests | 70–80 |
| PR2 ([#20057](https://github.com/apache/druid/pull/20057)) | Metadata, segment metadata, catalog, Guice, lookup, and server indexing tests | 70–80 |
| PR3 ([#20061](https://github.com/apache/druid/pull/20061)) | Remaining segment tests, RPC, Curator, discovery, messages, and general initialization | 65–75 |
| PR4 ([#20059](https://github.com/apache/druid/pull/20059)) | Coordinator tests, duties, rules, loading, balancing, simulations, and compaction | 70–80 |
| PR5 ([#20058](https://github.com/apache/druid/pull/20058)) | Server coordination, metrics, logging, security, scheduling, broker, audit, and remaining server tests | 70–80 |
| PR6 ([#20062](https://github.com/apache/druid/pull/20062)) | `services`, `embedded-tests`, HTTP, and Jetty/server initialization tests | 70–80 |

The shared-fixture portion of PR1 is now small because the processing helpers are already on `master`. PR1 only needs to address the remaining server cross-module cases:

- make `QueryStackTests` and its consumers use `GroupByQueryRunnerTestHelper` instead of the JUnit 4-based `GroupByQueryRunnerTest`;
- preserve or improve the Jupiter-compatible Derby connector while retaining the legacy rule adapter; and
- migrate the client/query tests that consume these fixtures.

### Dependency structure

Under this six-PR plan, PR1 is the only prerequisite. PR2–PR6 are independent siblings and should each branch directly from `master` after PR1 merges. They must not be stacked on one another.

```text
PR1: shared server fixtures + first server batch
                         |
                 merged into master
             /      /      |      \\      \\
           PR2    PR3     PR4     PR5     PR6
             \\      \\      |      /      /
                    final cleanup
```

If all six migration PRs must be completely independent, the remaining shared server-fixture changes should first be extracted into a separate small foundation PR. Then PR1–PR6 can all branch from the updated `master`, at the cost of one additional prerequisite PR.

JUnit 4, Vintage, migration-support, and Hamcrest dependencies must remain until the remaining processing Batch 4 ([#19980](https://github.com/apache/druid/pull/19980)), MSQ, indexing-service, server, services, and embedded tests are migrated. The final cleanup PR will then remove those dependencies, the obsolete OpenRewrite migration rule, and `dev/validate-junit5-usage` after a repository-wide audit and static-check run.

**Comment by FrankChen021** [comment]
migration completed via above PRs.
